create testable package in subdir

pawelru opened the pull request:
Implementation details:
- Create a package in `/package` subdir.
  - There is clear redundancy between `./DESCRIPTION` (i.e. book DESC file) and `./package/DESCRIPTION` when it comes to the dependency specification. It is needed to specify it on package level - otherwise R CMD CHECK complains. It's very likely that it's not necessarily needed on the book level but I would have to read deployment logic first.
- The book files and dirs is symlinked into `./package/testthat/_book/`
  - This is a limitation of this idea as any significant changes to the book structure needs to be re-symlinked again
  - The book as a whole cannot be symlinked as a single dir (i.e. `ln -s . ./package/testthat/_book` because R CMD BUILD does not interpret symlink resulting in recursive loop creating infinite file paths: `./package/testthat/_book/package/testthat/_book/package/...`)
- For each of the article
  - create envir upfront
  - for each output: append results to that environment
  - dump envir to rds file in `./package/testthat/_data/`
- Testing strategy:
  - as a part of setup: render the book stored in `./package/testthat/_book/` and expect files in `./package/testthat/_data/`
  - for each article: load appropriate rds file and for each output: do snapshot testing

As a result:
- we can execute R CMD CHECK and R CMD TEST on the `./package/` subdir and we can include that in our CI - in particular: revdepcheck
- symlinks makes the updates to the book automatically incorporated in the package

Known limitations (might be resolved):
- the need of re-symlink on book structure changes
- redundancy of deps specifications between book and package

Some other ideas (not yet implemented nor declined):
- quarto supports profiles ([docs](https://quarto.org/docs/projects/profiles.html)) and I am thinking now if during testing we can have (limited) book render that will skip stuff like teal image snapshots, renv snapshots that takes considerable amount of time.

shajoezhu commented:
can we use softlinks for description, .rbuildignore files etc

shajoezhu commented:
> > for description
> 
> Unfortunately not. Tried this path already. For book it has `Type: Book` that blocks any R CMD * as found by Davide.
> 
> > .rbuildignore etc.
> 
> Probably yes but that's not a big overhead to be honest.

hey @pawelru , I agree, the overhead is not big. but if we keep two copies, we still need to maintain two copies, and this will be  duplicated work for maintaining. thinking out loud, if we don't keep a hard copy of DESCRIPTION file, but when running r cmd check on in cicd, use sed "s/" and modify the description file, and create it on the fly. would this work?

pawelru commented:
Oh sorry I previously meant not a big overhead for all ignore type of the files. For DESCRIPTION you are definitely right. I share the same concerns. Haven't yet found out a clever way to address that.

pawelru commented:
I am thinking now that in order to address two limitations we would have to have a package as a top-level and book as a child or alternatively there will be two top level dirs: package and book. This would allow to create symlink to the whole book directory as opposed to currently book subdirs and subfiles individually. 
```
├── tlg.catalog
│   ├── **book** - optional symlink for convinience
│   ├── R
│   │   └── *.R
│   ├── DESCRIPTION
│   ├── NAMESPACE
│   ├── tests
│   │   ├── testthat
│   │   │   ├── **book**
│   │   │   │   └── (book files)
│   │   │   ├── _data
│   │   │   ├── _snaps
│   │   │   └── test-*.R
│   │   └── testthat.R
```
Or:
```
├── tlg.catalog
│   ├── book
│   │   └── (book files)
│   ├── package
│   │   ├── R
│   │   │   └── *.R
│   │   ├── DESCRIPTION
│   │   ├── NAMESPACE
│   │   ├── tests
│   │   │   ├── testthat
│   │   │   │   ├── _book -> a symlink to the toplevel book
│   │   │   │   ├── _data
│   │   │   │   ├── _snaps
│   │   │   │   └── test-*.R
```

I think I like the latter more. Let me experiment with it a little.

pawelru commented:
So I have decided to go towards the latter, i.e. separate top-level dir for a package and a book. I think it's getting nicer and nicer. Please have a look and tell me your thoughts on this.

Some few enhancements:
- symlink to the whole book allowing a lot of freedom in the future book development
- render within tests speed-up achieved by disabling renv spanshotting
- remove dependency specification from DESC in book and leave it only in package (i.e. where it's actually being used during R CMD CHECK)

As a result - I have addressed all the limitations I have identified previously.

New challenges:
- As anticipated, there are some challenges in the workflows. I hope that with the help of IDR this will be easy to pick up:
  - currently, package level workflows does not enable path argument (i.e. when package is in subdir) - to be resolved here: https://github.com/insightsengineering/idr-tasks/issues/595. I have implemented a workaround by keeping DESC and NEWS in top-level and symlink it also into the `/book`
  - there is a new need for a new logic to copy version and date between `/book` and `/package`
- this might interfeer in the release cycle of TLGCat but I don't yet have all the details (vide: upcoming research on stable/devel version). My assumption is that the package won't be a release at all.

pawelru commented on .github/workflows/check.yml:
```diff
@@ -34,6 +34,33 @@ jobs:
     uses: insightsengineering/r.pkg.template/.github/workflows/version-bump.yaml@main
     secrets:
       REPO_GITHUB_TOKEN: ${{ secrets.REPO_GITHUB_TOKEN }}
+  copy_vbump:
```
@cicdguy can you please have a look?
It's a new need regardless of path parameter to workflows. The goal is to update DESC in `/book` (incl. vbump, date etc.) first in standard `vbump` workflow and then copy it all into the `/package/DESC`. The goal is to keep consistency between versioning of those two so as to avoid confusion.
We might also consider two vbumps in two subdirs separately. Maybe it's easier to maintain? 🤔

pawelru replied:
Thinking about that a little bit more, we only need it on package level (so that it's present in CI that tlg.catalog.pkg@1.2.3.9004 failed). Having this on book level doesn't provide any value. Let me adjust that PR then

pawelru replied:
I will just keep one in package not to confuse people. When within a book, one might want to add book-specific entries or (even worse) remove package-specific fields. Let's go with simpler at the beggining.

Melkiades commented on book/si.qmd:
````diff
@@ -13,11 +13,12 @@ sessionInfo()
 ```{r snapshot}
 #| code-fold: show
 #| renv.ignore: TRUE
+knitr::opts_chunk$set(eval = isTRUE(testthat::is_checking() || testthat::is_testing()))
````
Nice!!!

Melkiades reviewed (commented):
I think already "as is" looks great. Very effective division between book and package. Ideally we will never have to touch the package part again (beside deps).

Maybe also @edelarua and @ayogasekaram can take a look about this version

pawelru commented on .github/workflows/check.yml:
```diff
@@ -20,6 +20,22 @@ jobs:
+        checking R code for possible problems ... NOTE
+        checking examples ... NOTE
+        checking Rd line widths ... NOTE
+    # need a new path parameter set to "./package"
```
FYI @cicdguy I have added r-cmd-check workflow but we are obviously miss path parameter. From what I noticed it's needed also for version-bump and version-check. It's nice to have for spelling but here we are not going to use it.

walkowif replied:
The feature to specify a subdirectory in the repository, where the package is located, has just been merged to `r.pkg.template`. The new input called `package-subdirectory` is available in workflows in which the location of the package within the repository is relevant.

pawelru replied:
Yes please. Thanks!

pawelru replied:
Please have a look at 37cc362a3d6cc6a7f29b0d69caeae437d725bdfd - in particular: removed workflows to be added back as well as changed render call which might be very useful when debugging (I expect R CMD CHECK to take a while since we are rendering the whole book)

walkowif replied:
I think the R CMD check is getting skipped because the [package name in `DESCRIPTION`](https://github.com/insightsengineering/tlg-catalog/actions/runs/5508960962/jobs/10040983744#step:18:25) is different than the [repository name](https://github.com/insightsengineering/tlg-catalog/actions/runs/5508960962/jobs/10040983744#step:20:49).

pawelru replied:
> I think the R CMD check is getting skipped because the [package name in `DESCRIPTION`](https://github.com/insightsengineering/tlg-catalog/actions/runs/5508960962/jobs/10040983744#step:18:25) is different than the [repository name](https://github.com/insightsengineering/tlg-catalog/actions/runs/5508960962/jobs/10040983744#step:20:49).

This is quite uncommon but still it should be allowed. The most popular example is https://github.com/vubiostat/r-yaml that has `yaml` package inside. IMO: build step should return a path to .tar.gz file for check step to be picked up.

pawelru replied:
I have changed package name to `tlg-catalog` to mach repo name but then R CMD BUILD failed as follows:
```
── R CMD build ──────────────────────────────────────────
   checking for file ‘/Users/ruckip/Documents/repo/gh/ins✔  checking for file ‘/Users/ruckip/Documents/repo/gh/insightsengineering/tlg-catalog/package/DESCRIPTION’
─  preparing ‘tlg-catalog’: (922ms)
E  checking DESCRIPTION meta-information ...
   Malformed package name
   
   See section 'The DESCRIPTION file' in the 'Writing R Extensions'
   manual.
Error in `(function (command = NULL, args = character(), error_on_status = TRUE, …`:
! System command 'R' failed
```

This workaround is not possible.

walkowif replied:
Ok, I added the support to R CMD check workflow for packages with names different than the repository name.
Now the workflow run looks like [this](https://github.com/insightsengineering/tlg-catalog/actions/runs/5520600244/jobs/10067455146).

pawelru replied:
actually the `staged_dependencies.yaml` file is at the repo root level so this should be picked up...

pawelru replied:
ah no... the staged-deps step is actually a part of check workflow so it looks for that config file in the package subdir. let me move it

walkowif replied:
I think we encountered this issue: https://github.com/openpharma/staged.dependencies/issues/172 🙂

walkowif replied:
I don't think so because it's already set by [this input](https://github.com/insightsengineering/tlg-catalog/actions/runs/5520740988/jobs/10067799927#step:15:3) via [staged-dependencies-action](https://github.com/insightsengineering/staged-dependencies-action/blob/main/action.yml#L92).

Melkiades commented:
I think this has priority @edelarua @pawelru @cicdguy. It would make reviewing by comparison with other results much faster and more reliable

pawelru commented:
Let me see if this can be merged without CI in place (to be delieved later on after merge). Will do that next week

pawelru commented:
I think I have found even nicer way to store outputs for testing via custom knitr hook. Right now you have to modify code chunk header as opposed to add a new one afterwards. I think it's much cleaner. The syntax is as follows:
`{r <label>, <other options>, test = list(<target object name> = "<actual variable name>")}`, e.g. `{r variant1, test = list(result_v1 = "result")}`

I have also removed `opts_chunk` and `opts_current` as this was not really working. Luckily, chunks header supports R expressions as chunk option.

Current state should be fully functional.

Melkiades commented on package/tests/testthat/helper.R:
```diff
@@ -0,0 +1 @@
+test_data_path <- testthat::test_path("_data")
```
Do we need this and the same in `setup.R`?

pawelru replied:
I think not - that would not be DRY. I move it to helper for interactive debugging. 

> Helper files live in tests/testtthat/, start with helper, and end with .r or .R. They are sourced by devtools::load_all() (so they’re available interactively when developing your packages)

> Setup files live in tests/testthat/, start with setup, and end with .r or .R. Typically there is only one setup file which, by convention, is tests/testthat/setup.R. Setup files are sourced by [test_check()](https://testthat.r-lib.org/reference/test_package.html) and friends (so that they’re available no matter how your tests are executed), but they are not sourced by devtools::load_all().

[source](https://testthat.r-lib.org/articles/special-files.html)

pawelru commented:
Woooho! It fiinally got green. 
Before we go with it - @cicdguy can you please have a look at the changes in the docs workflow?
